Add custom texture create function

BastiaanOlij opened the pull request:
This PR adds a method to `RenderingServer`, `texture_rd_create` that allows you to create a texture object using a RID relating to a texture created directly with the `RenderingDevice`. This in turn allows you to use that texture object to use the texture for materials and such.

For those who want to test, a little example project I wipped together:
[TestCustomTextures.zip](https://github.com/godotengine/godot/files/12102369/TestCustomTextures.zip)

Looks way better animated but it implements an old fashion ripple effect as a compute shader:
![image](https://github.com/godotengine/godot/assets/1945449/91d2e20c-6b3f-4745-ae45-9dcec5f84c1b)

BastiaanOlij committed:
Add custom texture create function

reduz commented on servers/rendering/renderer_rd/storage_rd/texture_storage.cpp:
```diff
@@ -1429,6 +1429,60 @@ Size2 TextureStorage::texture_size_with_proxy(RID p_proxy) {
 	return texture_2d_get_size(p_proxy);
 }
 
+void TextureStorage::texture_rd_initialize(RID p_texture, const RID &p_rd_texture) {
```
I wonder, should this RID be also freed when the texture is freed, or should it be freed manually from RenderingDevice?
What do you think?

BastiaanOlij replied:
I'm a strong believe over the design pattern that the entity that creates a resource should also free it. So if I create the texture on RenderingDevice I should also free it.

One thing that we could do here, and that solves some other problems too, is that this always creates a new shared texture on the texture given to this function. Then destroying this texture destroys the texture view, but you're still left with the original texture to destroy on the rendering device.

BastiaanOlij replied:
Another question around this that could make things difficult for us. In theory someone could create a texture object through the `RenderingServer`, call `texture_get_rd_texture` to get the RID to the RD texture, and then call `texture_rd_create` to create a new texture object for that RD texture.
There is however no dependency between the original texture and the new texture meaning that if the original texture is freed, our new texture object will contain an invalid RID.

BastiaanOlij replied:
Come to think of it, we should expose `texture_get_rd_texture` to users as well. So far this has only been used internally for XR I think.

clayjohn replied:
I guess we can just throw an error in ``texture_rd_create()`` if it is called using an RD texture that already belongs to a RS::Texture. But I guess that only works if we track dependencies on the RD side, the way we do on the RS side

BastiaanOlij replied:
Yeah we currently don't track those dependencies and I feel we already have a big memory impact with the way we track those atm.

clayjohn replied:
I guess this is fixed by using a shared texture instead of creating a new texture right?

BastiaanOlij replied:
Yeah I think it works this way now

clayjohn commented:
CC @joemarshall

BastiaanOlij commented on drivers/vulkan/rendering_device_vulkan.cpp:
```diff
@@ -2874,6 +2874,32 @@ bool RenderingDeviceVulkan::texture_is_valid(RID p_texture) {
+
+	TextureFormat tf;
+
+	// Q? I guess we need to adjust this if this is a shared texture
```
This may not be the right thing to do here but we may at least need to add a `texture_get_owner` method when we're dealing with a shared texture.

BastiaanOlij commented on servers/rendering/renderer_rd/storage_rd/texture_storage.cpp:
```diff
@@ -1429,6 +1429,78 @@ Size2 TextureStorage::texture_size_with_proxy(RID p_proxy) {
+	ERR_FAIL_COND(!RD::get_singleton()->texture_is_valid(p_rd_texture));
+
+	// TODO : investigate if we can support this, will need to be able to obtain the order and obtain the slice info
+	ERR_FAIL_COND_MSG(!RD::get_singleton()->texture_is_shared(p_rd_texture), "Please create the texture object using the original texture");
```
So for now I've disabled this seeing we're going to create shared textures from this. May change this depending on our discussion around whether we're going to take ownership of the resource or not.

clayjohn replied:
I don't think we should support creating RS::textures from shared textures. The relationships will get every complex if we can:
Create a texture: A
create a view from texture A: B
Create a texture from B: C
create a view from C: D

Although maybe it will be possible to support in the future.

I think for now the error is the best approach

BastiaanOlij replied:
Cool, and once you have the texture you can always use the RS methods for creating a shared texture, but honestly I can't really see a use case here. The RS::texture will be used primarily for display with shaders and shaders can already select the required layer and/or mipmap level. Generally slices are created to render to a specific layer and you would keep that on the RD side.

BastiaanOlij commented on servers/rendering/renderer_rd/storage_rd/texture_storage.cpp:
```diff
@@ -1429,6 +1429,78 @@ Size2 TextureStorage::texture_size_with_proxy(RID p_proxy) {
+	texture.validated_format = texture.format; // ??
+
+	RD::TextureView rd_view;
+	rd_view.format_override = imfmt.rd_format == tf.format ? RD::DATA_FORMAT_MAX : imfmt.rd_format;
```
If the user changes the swizzle data when creating their texture, we have no way to obtain that information, we don't store it. So for now I'm assuming the swizzle info was used as we determine it in `_validate_texture_format`.

We may want to enhance the rendering device to record the swizzle information when creating the texture  (slice) so we can retrieve it. Then we can also remove the swizzle data from `_texture_format_from_rd`

BastiaanOlij replied:
It is my understanding the `TextureView` overrides but I must be honest that this is an area I haven't fully figured out yet because I have no real world experience with it. Logic suggest that most users will stick to RGBA formats and will never deviate from the required swizzle so it may very well be a non issue in practice.

BastiaanOlij commented on servers/rendering/renderer_rd/storage_rd/texture_storage.cpp:
```diff
@@ -1429,6 +1429,78 @@ Size2 TextureStorage::texture_size_with_proxy(RID p_proxy) {
+		texture.rd_texture_srgb = RD::get_singleton()->texture_create_shared(rd_view, p_rd_texture);
+	}
+
+	// TODO figure out what to do with slices
```
I have no idea if we need to populate this data...

BastiaanOlij commented on servers/rendering/renderer_rd/storage_rd/texture_storage.cpp:
```diff
@@ -1921,6 +1993,362 @@ Ref<Image> TextureStorage::_validate_texture_format(const Ref<Image> &p_image, T
 
+void TextureStorage::_texture_format_from_rd(RD::DataFormat p_rd_format, TextureFromRDFormat &r_format) {
+	switch (p_rd_format) {
+		case RD::DATA_FORMAT_R8_UNORM: {
```
Do we want to support more formats like the BGR ones here?

clayjohn replied:
We should support all the formats that we currently support in the RS

BastiaanOlij replied:
Yeah that is what this code is based on, I basically took `_validate_texture_format` and reversed it.
However RD supports many more formats that in theory could be used here.

But I suggest we keep it as is, and if anyone ever runs into a real live use case for using other formats, we'll add the required entries.

BastiaanOlij commented on scene/resources/texture.cpp:
```diff
@@ -3547,3 +3547,82 @@ PlaceholderTextureLayered::~PlaceholderTextureLayered() {
+}
+
+bool Texture2DRD::has_alpha() const {
+	return false;
```
How important is reporting this correctly? The only way I can think of doing this is checking each and every `DATA_FORMAT_...` and seeing which ones have an alpha channel. I guess we could add a function for this to `RD::TextureFormat`.

reduz commented on scene/resources/texture.h:
```diff
@@ -1093,4 +1093,29 @@ class PlaceholderCubemapArray : public PlaceholderTextureLayered {
 			PlaceholderTextureLayered(LAYERED_TYPE_CUBEMAP_ARRAY) {}
 };
 
+class Texture2DRD : public Texture2D {
```
I think these classes should probably be put in the servers/rendering/renderer_rd, so if this back-end is disabled, these textures are disabled too.

BastiaanOlij replied:
That makes sense

reduz reviewed (commented):
Looks good, just the comments noted (also check comment about freeing the RD RID above).

sakrel commented:
Does this play nice when RenderingServer is in multi-threaded mode?
Imagine you create a texture with RenderingDevice, you then
1) access that texture with RenderingServer
2) and after that access the texture with RenderingDevice again. 

It's my understanding that when RenderingServer is in multi-threaded mode, RenderingServer commands are queued and then executed at a later point on another thread. How do you make sure that RenderingServer commands in step 1) execute before step 2)?

clayjohn commented:
@sakrel Ultimately the RenderingServer is calling into the RenderingDevice to handle the texture. As all the RenderingServer deals with is the RID. Since the RenderingDevice is thread safe, everything should just work shouldn't it?

I should say "since the RenderingDevice is thread safe, _in theory_" Since we have a bit of work to do to make it more thread safe and efficient. Right now we just throw mutexes at the beginning of every method we want to be thread safe

sakrel commented:
> @sakrel Ultimately the RenderingServer is calling into the RenderingDevice to handle the texture. As all the RenderingServer deals with is the RID. Since the RenderingDevice is thread safe, everything should just work shouldn't it?

Yes, the RenderingDevice itself should be fine. It's just that when the RenderingServer is multi-threaded, it's possible for function calls to be reordered. Take this code for example:
```
	var thread_model: int = ProjectSettings.get_setting("rendering/driver/threads/thread_model");
	if thread_model == 0:
		print("Thread Model: ", "Single-Unsafe");
	elif thread_model == 1:
		print("Thread Model: ", "Single-Safe");
	elif thread_model == 2:
		print("Thread Model: ", "Multi-Threaded");
		
	
	var rd := RenderingServer.get_rendering_device();
	var format := RDTextureFormat.new();
	format.format = RenderingDevice.DATA_FORMAT_R8G8B8A8_UNORM;
	format.texture_type = RenderingDevice.TEXTURE_TYPE_2D;
	format.usage_bits = RenderingDevice.TEXTURE_USAGE_SAMPLING_BIT | RenderingDevice.TEXTURE_USAGE_CAN_UPDATE_BIT | RenderingDevice.TEXTURE_USAGE_CAN_COPY_FROM_BIT | RenderingDevice.TEXTURE_USAGE_CAN_COPY_TO_BIT;
	format.width = 1;
	format.height = 1;
	var rd_texture := rd.texture_create(format, RDTextureView.new());
	var texture := RenderingServer.texture_rd_create(rd_texture);
		
	rd.texture_clear(rd_texture, Color.WHITE, 0, 1, 0, 1);
	
	var image :=  Image.create(format.width, format.height, false, Image.FORMAT_RGBA8);
	image.fill(Color.BLACK);
	RenderingServer.texture_2d_update(texture, image, 0);

	var data := rd.texture_get_data(rd_texture, 0);
	print("texture_get_data Color: ", data);
```
So, first we use `RenderingDevice::texture_clear` to clear the RenderingDevice texture to white, then we use `RenderingServer::texture_2d_update` to clear the texture to black.

With thread model `Single-Safe`, I get the following results:
>Thread Model: Single-Safe
texture_get_data Color: [0, 0, 0, 255]

As expected, the texture color is black.
When I run the same code with thread model `Multi-Threaded` the texture color is white:
>Thread Model: Multi-Threaded
texture_get_data Color: [255, 255, 255, 255]

I think this happens because the engine doesn't execute `RenderingServer.texture_2d_update(texture, image, 0);` immediately. Instead, the `RenderingServer.texture_2d_update` command is enqueued and executed at a later point by the RenderingServer thread.

joemarshall commented:
Like this pr, made a few comments.

About threading, nothing quite works right now - I've got a work in progress to threadify renderingdevice - there's a few API changes needed which is annoying - most obviously draw_list_end which doesn't take a draw list ID, along with the various message things which underlying are using a draw list but without taking an id.

BastiaanOlij commented on servers/rendering/renderer_rd/storage_rd/texture_storage.cpp:
```diff
@@ -1429,6 +1429,78 @@ Size2 TextureStorage::texture_size_with_proxy(RID p_proxy) {
+	RD::TextureFormat tf = RD::get_singleton()->texture_get_format(p_rd_texture);
+	ERR_FAIL_COND(!(tf.usage_bits & RD::TEXTURE_USAGE_SAMPLING_BIT));
+
+	TextureFromRDFormat imfmt;
+	_texture_format_from_rd(tf.format, imfmt);
```
I did it this way seeing `_validate_texture_format` already did it this way, but I'm not married to it :P

BastiaanOlij commented:
@sakrel I think that use case would always result in something unexpected, at some point we need draw a line whether users are doing something logical or are just trying to trip up the system. My only "fear" in this is that documentation may not be clear and people take an overcomplicated approach because they are looking at two completely unrelated examples and trying to merge them.

The whole premise is that you will only work with `RenderingDevice` directly when you want to create your own systems for rendering too textures, be it compute shaders or through framebuffers where you're rendering stuff that viewports aren't suitable for.

So it will always be through `RenderingDevice` that you are populating the texture. The `RenderingServer` texture is purely used for consuming that data, being able to display the contents of the texture or use it in materials.

Having a scenario where both RD and RS are writing to the texture and doing things that contradict each other, you get into trouble. For instance for your example `texture_2d_update`, you're actually not rendering to the RD texture, you're replacing it with a new RD texture object created from the image you supplied.

The docs must be clear on this, you can't interchange `texture_2d_...`, `texture_3d_...`, `texture_rd_...` functions.

BastiaanOlij commented:
@clayjohn don't they all work the same? The only thing `texture_2d_create` does outside of the thread is reserve the RID for the texture, it doesn't actually populate the texture structure nor create anything in the rendering servers until `texture_2d_instantiate` is called, which has the same thread protection code as `texture_2d_update`.

But I was partly wrong in what I said, you can populate and rd texture with `texture_2d_update`. But assuming `texture_rd_create` and `texture_2d_update` are called from the same thread, they will either both be executed right away (if we're on the main thread) or both queued (if we're not on the main thread). It's only the reserve of the RID that happens right away which is important as that is our denominator.

clayjohn commented:
@BastiaanOlij Take a look at the code block I posted above. ``texture_2d_create()`` calls ``texture_2d_instantiate()`` immediately when using the RD renderer. It is only queued when using the GLES3 renderer

BastiaanOlij commented on servers/register_server_types.cpp:
```diff
@@ -247,6 +248,13 @@ void register_server_types() {
 	GDREGISTER_CLASS(RDShaderFile);
 	GDREGISTER_CLASS(RDPipelineSpecializationConstant);
 
+	GDREGISTER_CLASS(Texture2DRD);
```
We don't have a place to register `renderer_rd` classes in the `renderer_rd` folder. If the renderer is excluded this will fail. I'm also wondering why we are including `rendering_device` and declaring its binding classes outside of `renderer_rd`. 

One thing we discussed on chat was that `renderer_rd` should really be a module but that is a big change that we don't want to do right now.

I'm guessing I should just put an `#ifdef` around this block?

clayjohn replied:
I don't think we have a compile time macro to remove the RD renderer. So I think this is fine for now

BastiaanOlij commented:
> @BastiaanOlij Take a look at the code block I posted above. `texture_2d_create()` calls `texture_2d_instantiate()` immediately when using the RD renderer. It is only queued when using the GLES3 renderer

Ah because of the `can_create_resources_async`?

Hmm, so if I understand correctly, the problem is that if you're **not** on the main thread, and you call `texture_2d_update`, that is now queued. When you then try to access the RD texture, depending on what `texture_2d_update` does, you may be getting the value as it was before the change is applied, not after.

I think we have that problem in many places, basically an object should be marked `pending` if it's added stuff to the command queue and any getter should yield the thread to let anything on the command queue be processed by the main thread before continuing.

joemarshall commented:
Currently when you call texture_get_data, there's a call to flush, which flushes all draw calls, so in theory it should get everything that is drawn on whatever thread. Although it is a total pipeline stall which means that anything using texture_get_data is a bad idea anyway; that needs deprecating and replacing with async texture fetchers some point soon, or to return some kind of future-like object or something.

The problem being observed above isn't anything to do with renderingdevice though, it is to do with the renderingserver proxying stuff to the rendering thread using command queue. Which whilst it makes sense for the GL renderer, needs some consideration for vulkan as to whether it is still appropriate. 

My instinct on this is that:
a) RenderingServer should proxy things to the renderer thread so it is easy to understand and use. This should include everything except for resource creation calls I think, although I don't quite know how much performance benefits creating resources on multiple threads actually gives, and as you see above, it creates confusion at points.

b) renderingdevice should support writing draw buffers, updating textures etc. and everything on multiple threads. 

c) Consistency within threads should exist, i.e. if you call something to draw into a texture on a thread, then call something to read that texture on the same thread, you should get the right answer.

d) The order of GPU operations between threads should not necessarily be guaranteed - that way we only have to flush everything at the end of frame, rather than introducing dependencies between threads.

e) We should really really strongly discourage combination of renderingserver and renderingdevice calls like the above, except where absolutely required during the process of transferring resources from one to the other. It is totally asking for trouble...

What this basically means from a vulkan point of view is that there is a separate vulkan command list for each thread. All command lists are transferred on frame end as they are currently. As we primarily use one queue for drawing, texture uploads etc, within thread consistency should be fine. texture_get_data etc. should firstly be discouraged, but secondly should push all pending commands (or maybe just the ones pending on that thread?) including the commands to copy data back, protected by appropriate barriers, then wait on a fence, as opposed to the current situation which is that it waits until everything pending is processed and the GPU is idle. Which as I mention above is *BAD*.

joemarshall commented:
More specifically relating to my comment above

This kind of combination of things - rendering device creating something, then a renderingserver texture wrapping it, that should be fine. It should be documented that texture_rd_create calls immediately and is guaranteed to be done on return.

```
var rd_texture := rd.texture_create(format, RDTextureView.new());
var texture := RenderingServer.texture_rd_create(rd_texture);
```

But the this kind of stuff - updating a texture using renderingserver, then expecting it to be updated in a renderingdevice call, is not fine. We should be clear that renderingserver dispatches work which may or may not be done immediately (and isn't in the case of a render thread). If you call other renderingserver calls they should wait for consistent state, but any attempt to make renderingdevice calls wait for consistent state risks limiting the potential for performance gains that multithreaded rendering can give us.

```
RenderingServer.texture_2d_update(texture, image, 0);
var data := rd.texture_get_data(rd_texture, 0);
```

joemarshall commented:
> Hmm, so if I understand correctly, the problem is that if you're **not** on the main thread, and you call `texture_2d_update`, that is now queued. When you then try to access the RD texture, depending on what `texture_2d_update` does, you may be getting the value as it was before the change is applied, not after.

Incidentally, what is the **main** thread here depends on the rendering setting. When multithreaded rendering is turned on, godotscript is never running on the rendering thread, so if renderingserver calls are queued, renderingdevice calls will do their thing before the renderingserver call gets a chance to run the queued calls. All the more reason why the renderingdevice calls should never be mixed with renderingserver except in very specific circumstances.

BastiaanOlij commented:
@joemarshall you make a good point that we need a way to return a future, I don't think we have that concept in Godot yet but that really is the only way to deal with scenarios like that.

We have similar issues with getting mesh data which we do in the IDE (something I ran into by accident because it fails on Android which does not seem to have data ready on flush).

BastiaanOlij commented:
> Incidentally, what is the main thread here depends on the rendering setting. When multithreaded rendering is turned on, godotscript is never running on the rendering thread, so if renderingserver calls are queued, renderingdevice calls will do their thing before the renderingserver call gets a chance to run the queued calls. All the more reason why the renderingdevice calls should never be mixed with renderingserver except in very specific circumstances.

I'm not sure about this btw, I know this is so in Godot 3, but in Godot 4 we're using threading in a very different way. I've never been very confident about the way we're handling this in Godot 3 but I also have to admit I never investigated it throughly.

The only reason you'd want rendering to run on another thread is that you want to start processing the next frame, but you really can't do that until after you've made a snapshot of current instance data, which is something we do while looping through our viewports.

Our current approach of using threading in the rendering server itself controlled from the main thread seems a much better approach, where we efficiently handle the current frame, build the command buffers, and then sent them off to the GPU, the GPU can then do its work in parallel with us processing the next frame (which now also makes use of multi threading to paralyze a lot of the work). 

Having a separate rendering thread seems like little win while complicating matter and requiring some serious synching anyway.

BastiaanOlij commented:
I've moved this out of draft and added an example project so people can start playing with it.

I do not know what I did that caused the CI to fail on documentation on texture classes this PR doesn't touch.

clayjohn commented on servers/rendering/renderer_rd/scene/texture_rd.cpp:
```diff
@@ -0,0 +1,340 @@
+
+RID Texture2DRD::get_rid() const {
+	if (texture_rid.is_null()) {
+		//we are in trouble, create something temporary
```
```suggestion
		// We are in trouble, create something temporary.
```

clayjohn commented on servers/rendering/renderer_rd/scene/texture_rd.cpp:
```diff
@@ -0,0 +1,340 @@
+
+		size.width = tf.width;
+		size.height = tf.height;
+		// TODO add check to see which formats have alpha... Is this even important?
```
Looks like most texture types just return false. This was likely used a long time ago, but isn't functional anymore. I would leave it as is (with comment removed) and then we can add a check if needed

clayjohn commented on servers/rendering/renderer_rd/storage_rd/texture_storage.cpp:
```diff
@@ -1429,6 +1436,78 @@ Size2 TextureStorage::texture_size_with_proxy(RID p_proxy) {
+	Texture texture;
+
+	switch (tf.texture_type) {
+		// Q? What to do with _1D and cube textures?
```
RS doesn't do anything with 1D textures, so we can ignore them for now. But this needs to support CUBEMAP and CUBEMAP_ARRAY as they are supported by ``texture_2d_layered_initialize``

BastiaanOlij replied:
the problem is that on the RD side we don't identify them as such, it's just a 2D texture with 6 layers for CUBEMAP and a multiple of 6 for CUBEMAP_ARRAY. Take a look at the TextureCubemapRD implementation I did.

BastiaanOlij replied:
@clayjohn `texture_2d_layered_create` isn't used here. You mean adding the parameter to `texture_rd_create`/`texture_rd_initialize` ?

BastiaanOlij replied:
@clayjohn let me know what you think of the changes I made

clayjohn replied:
Looks good! You can delete the comment now
```suggestion
```

clayjohn commented on doc/classes/RenderingDevice.xml:
```diff
@@ -752,6 +752,13 @@
+			<return type="RDTextureFormat" />
+			<param index="0" name="texture" type="RID" />
+			<description>
+				Returns the format data used to create this texture with.
```
```suggestion
				Returns the data format used to create this texture.
```

clayjohn commented on doc/classes/TextureLayeredRD.xml:
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+<class name="TextureLayeredRD" inherits="TextureLayered" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="../class.xsd">
+	<brief_description>
+		Texture Array that is bound to a texture created on the [RenderingDevice].
```
Abstract base class for layered texture RD types.

clayjohn commented on doc/classes/TextureLayeredRD.xml:
```diff
@@ -0,0 +1,16 @@
+		Texture Array that is bound to a texture created on the [RenderingDevice].
+	</brief_description>
+	<description>
+		This texture class allows you to use an array texture created directly on the [RenderingDevice] as a texture for materials, meshes, etc.
```
Base class for [Texture2DArrayRD], [TextureCubemapRD] and [TextureCubemapArrayRD]. Cannot be used directly, but contains all the functions necessary for accessing the derived resource types.

clayjohn reviewed (commented):
Amazing work! I think this is nearly ready to go. I left a few comments and suggestions. The biggest missing work item is to add support for Cubemaps and Cubemap Arrays on the RS side. Other than that I think this is ready to go. 

I tested it locally with your attached demo and it works great. We should get your demo uploaded to the demos repo as soon as we merge this

BastiaanOlij commented:
Demo has now been added to the projects: https://github.com/godotengine/godot-demo-projects/pull/938

BastiaanOlij commented on doc/classes/AnimatedTexture.xml:
```diff
@@ -59,7 +59,6 @@
 		<member name="pause" type="bool" setter="set_pause" getter="get_pause" default="false">
 			If [code]true[/code], the animation will pause where it currently is (i.e. at [member current_frame]). The animation will continue from where it was paused when changing this property to [code]false[/code].
 		</member>
-		<member name="resource_local_to_scene" type="bool" setter="set_local_to_scene" getter="is_local_to_scene" overrides="Resource" default="false" />
```
I don't know why doctool insists on removing these, it doesn't seem related to this PR but a left over from the PR that broke the texture classes up into separate files (seems to be a recent change)?

clayjohn replied:
Have you tried rebasing?

BastiaanOlij replied:
yeah on master from a few hours ago :)

clayjohn replied:
As discussed on RC, this appears to happen as a result of defining the RD texture types in the register_server_classes code. Moving the registrations to scene makes this issue go away:
```
diff --git a/scene/register_scene_types.cpp b/scene/register_scene_types.cpp
index 1c2ed16e63..d86ae2ed28 100644
--- a/scene/register_scene_types.cpp
+++ b/scene/register_scene_types.cpp
@@ -226,6 +226,7 @@
 #include "scene/theme/theme_db.h"
 
 #include "scene/main/shader_globals_override.h"
+#include "servers/rendering/renderer_rd/scene/texture_rd.h"
 
 #ifndef _3D_DISABLED
 #include "scene/3d/area_3d.h"
@@ -887,6 +888,13 @@ void register_scene_types() {
        GDREGISTER_CLASS(PlaceholderCubemap);
        GDREGISTER_CLASS(PlaceholderCubemapArray);
 
+       GDREGISTER_CLASS(Texture2DRD);
+       GDREGISTER_ABSTRACT_CLASS(TextureLayeredRD);
+       GDREGISTER_CLASS(Texture2DArrayRD);
+       GDREGISTER_CLASS(TextureCubemapRD);
+       GDREGISTER_CLASS(TextureCubemapArrayRD);
+       GDREGISTER_CLASS(Texture3DRD);
+
        GDREGISTER_CLASS(Animation);
        GDREGISTER_CLASS(AnimationLibrary);
 
diff --git a/servers/register_server_types.cpp b/servers/register_server_types.cpp
index 01bb0bab7c..7cc9a82699 100644
--- a/servers/register_server_types.cpp
+++ b/servers/register_server_types.cpp
@@ -69,7 +69,6 @@
 #include "physics_server_3d.h"
 #include "physics_server_3d_wrap_mt.h"
 #include "rendering/renderer_compositor.h"
-#include "rendering/renderer_rd/scene/texture_rd.h"
 #include "rendering/rendering_device.h"
 #include "rendering/rendering_device_binds.h"
 #include "rendering_server.h"
@@ -248,13 +247,6 @@ void register_server_types() {
        GDREGISTER_CLASS(RDShaderFile);
        GDREGISTER_CLASS(RDPipelineSpecializationConstant);
 
-       GDREGISTER_CLASS(Texture2DRD);
-       GDREGISTER_ABSTRACT_CLASS(TextureLayeredRD);
-       GDREGISTER_CLASS(Texture2DArrayRD);
-       GDREGISTER_CLASS(TextureCubemapRD);
-       GDREGISTER_CLASS(TextureCubemapArrayRD);
-       GDREGISTER_CLASS(Texture3DRD);
-
        GDREGISTER_CLASS(CameraFeed);
```

BastiaanOlij replied:
We did end up deciding to move the classes into `scene` as well. But once we decide to move the renderer to a module we should add this back in.

clayjohn commented on servers/rendering_server.cpp:
```diff
@@ -1697,7 +1697,10 @@ void RenderingServer::_bind_methods() {
+	ClassDB::bind_method(D_METHOD("texture_get_format", "texture"), &RenderingServer::texture_get_format);
+
 	ClassDB::bind_method(D_METHOD("texture_set_force_redraw_if_visible", "texture", "enable"), &RenderingServer::texture_set_force_redraw_if_visible);
+	ClassDB::bind_method(D_METHOD("texture_rd_create", "rd_texture", "layer_type"), &RenderingServer::texture_rd_create);
```
This is missing a ``DEFVAL`` statement

clayjohn commented on servers/rendering/renderer_rd/storage_rd/texture_storage.cpp:
```diff
@@ -1429,6 +1436,78 @@ Size2 TextureStorage::texture_size_with_proxy(RID p_proxy) {
+			ERR_FAIL_COND(tf.array_layers != 1);
+			texture.type = TextureStorage::TYPE_2D;
+		} break;
+		case RD::TEXTURE_TYPE_2D_ARRAY: {
+			ERR_FAIL_COND(tf.array_layers == 1);
+			texture.type = TextureStorage::TYPE_LAYERED;
+			texture.layered_type = p_layer_type;
```
```suggestion
		case RD::TEXTURE_TYPE_2D_ARRAY: {
			// RenderingDevice doesn't distinguish between Array textures and Cube textures
			// this condition covers TextureArrays, TextureCube, and TextureCubeArray.
			ERR_FAIL_COND(tf.array_layers == 1);
			texture.type = TextureStorage::TYPE_LAYERED;
			texture.layered_type = p_layer_type;
```

krisp-idk commented:
hi! I tried following the example and ran into this error: `Trying to assign value of type 'Compressed2D' to a variable of type 'Texture2DRD'.`
I'm sure it's a user error since your example works fine, but I can't figure out what the difference in my code that causes the error is.
```if material:
	texture = material.get_shader_parameter("tex")
	if texture:
		texture.texture_rd_rid = output_tex```
`texture` is a Texture2DRD and `tex` is a sampler2D in the gdshader

critopadolf commented:
Can hint_screen_texture or a viewport be used as a Texture2drd?  

Viewport texture rid is seen as invalid by rd:

```C#
using Godot;
using System.Diagnostics;
public partial class minimum : SubViewport
{
    private RenderingDevice rd;
    public override void _Ready()
    {
        rd = RenderingServer.GetRenderingDevice();
        Debug.Assert(rd.TextureIsValid(this.GetTexture().GetRid()));
    }
}
```

and I'm not sure how to get hint_screen_texture's Rid (assuming there is one)

Thanks!

clayjohn commented:
> Can hint_screen_texture or a viewport be used as a Texture2drd?

The screen texture definitely cannot be used as a Texture2DRD. I don't think we allow using a Viewport's texture either. Texture2DRD is a texture resource, it fills a very different purpose from Viewports
